src/hotspot/share/utilities/nonblockingQueue.inline.hpp

```diff
@@ -42,14 +42,6 @@ void NonblockingQueue<T, next_ptr>::set_next(T& node, T* new_next) {
 template<typename T, T* volatile* (*next_ptr)(T&)>
 NonblockingQueue<T, next_ptr>::NonblockingQueue() : _head(nullptr), _tail(nullptr) {}
 
-#ifdef ASSERT
-template<typename T, T* volatile* (*next_ptr)(T&)>
-NonblockingQueue<T, next_ptr>::~NonblockingQueue() {
-  assert(_head == nullptr, "precondition");
-  assert(_tail == nullptr, "precondition");
-}
-#endif
```

Comment:
These assertions require the queue to be empty when the destructor is called. My static queue is destroyed at shutdown, and it may not be empty. I see no explicit problem with a queue not being empty when destroyed, I can only assume it was added to serve as an extra reminder for the specific use-case of g1dirtycardsetqueue which is the only existing user of NonblockingQueue. In my opinion it's better for individual queue owners to decide if they care about this.

Alternatives: I can modify NonblockingQueue to accept an argument (e.g. check_empty_at_shutdown) which is true for g1 card set and false for my usage. Or I can avoid NonblockingQueue entirely as there are a few other review concerns arising from its limitations.

Reply:
The asserts are there to catch memory leaks. It is the responsibility of the
owner of the NBQ to ensure it is empty before deleting it. (Probably that
should have been mentioned in a comment in the header.) This is a generic
requirement, and is not specific to G1DCQS (the sole previous client).

HotSpot code avoids variables with static storage duration for types with
non-trivial construction or destruction. (I thought this was discussed in the
style guide, but apparently not.  It's been discussed multiple times in
various reviews.)  That's what's happening here - there's no owning object for
the NBQ, and there's no explicit setup/teardown of it during VM initialization
and destruction.  That's the reason for hitting the asserts.

So no, don't add a configuration argument for NBQ.  I wouldn't approve that
either.

Reply:
Thanks. Is the 'leak' is relevant in this case since we're shutting down anyway? I tried draining the queue in before_exit but that doesn't seem to run with ctrl-c/SIGINT shutdowns.

> HotSpot code avoids variables with static storage duration for types with
non-trivial construction or destruction

Do you have a suggested alternative? Everything in symbol table seems to be static.

Reply:
I've filed a bug report against the style guide to provide guidance in this area.
https://bugs.openjdk.org/browse/JDK-8319242

We typically define a variable with a pointer type, and initialize it somewhere in the VM startup process.  So in
this case, something like

// declared in class
static TempSymbolDelayQueue* _cleanup_delay;

// in the .cpp
TempSymbolDelayQueue* SymbolHandleBase::_cleanup_delay = nullptr;
// and update uses accordingly.

Also declare and define an initialization function to initialize that variable, and call it somewhere in init.cpp,
probably in init_globals().  Use the style that exists there.  I expect if you don't put it early enough that you'll
reliably get an obvious crash from attempting to access a nullptr.

Reply:
As you suggest, in this case the "leak" doesn't matter because we're on the
way to shutdown. One might think the asserts could be conditionalized on that,
but so far as I know there isn't a reliable "on the way to shutdown" detector.
The idiom described above is what's done throughout HotSpot for initialization
(to ensure we have control of initialization order, among other things), and
we often don't worry about the teardown side of things.